### ISSUE TITLE
Documenting the processing steps for metadata

### ISSUE THREAD

**Issue by iherman** [w3c/csvw#193]
At the telco on [2015-02-04](http://www.w3.org/2015/02/04-csvw-minutes.html) the idea of documenting the processing steps came up. The [section on parsing](http://www.w3.org/TR/tabular-data-model/#parsing) is not the one, indeed, that section concentrates on the _parsing_ of a CSV file, provided the various dialect parameters are given, and it does not touch upon the exact processing steps to be followed to generate the right metadata, for example. I think such a section should be part of one of our documents, possibly as a normative section. 

I have made a first attempt below to describe the process; obviously to be refined. But that is my understanding of the current documents, and I may have of course misunderstood some things. But if I did, while doing the implementations, than maybe others will, too, which is all the more of a reason to document it...

TL;DR: the result is, unfortunately, complex. I would say: too complex, actually. 
# Processing steps
## If the process starts with the URL of a CSV file:
1. Get the raw CSV content from the file through an `HTTP GET` (or access it from the local file system). Store the Link Header from the return, too.
2. Get hold of the following metadata objects (see the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document): 
   - metadata supplied by the user of the implementation that is processing the tabular data (`UM`) 
   - metadata in a document linked to using a Link Header associated with the tabular data file (`LM`)
   - file-specific metadata in a document located based on the location of the tabular data file (`FM`)
   - directory-specific metadata in a document located based on the location of the tabular data file (`DM`)
3. Create a merge of these metadata `M1 = merge(UM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
4. If `M1` does not include a [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions), add a default (_this still has to be defined_)
5. Possibly modify the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` if the `;header=absent` is set in the `Content-type` of the returned Link Header (see the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#headers)) 
6. Parse the CSV data using the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` . An informal description on the effects of the various dialect parameter and the parsing itself is part of the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#parsing). 
7. Retrieve the metadata embedded within the tabular data file (`EM`) (see again the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document). Note that this specification does not define the syntax of that embedded metadata, if any, except if `EM` consists of a single header row (providing `title` values for the corresponding `column` objects)
8. Determine the default Metadata (`DM`) (_this still has to be formally defined and described_)
9. Create a merge of `M = merge(DM, UM, EM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
10. The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows according to the Dialect Description, with `M` as the final metadata.
11. If required by the user, use the raw data and `M` to generate either JSON or RDF
## If the process starts with the URL of a Metadata file:

T.B.D.
# Comments:
- step no. 9 is awkward. Of course, it can be optimized by storing `merge(LM,FM,DM)` separately, but the fact that the priority order places `UM` ahead of `EM` means that such a re-run becomes necessary
- same for DM (steps nos. 8 and 9): it is unclear where exactly in the process `DM` comes in and is also part of the re-merge.
- much of the complication comes, in my view, of the fact that step no. 6, i.e., the parsing itself, is _dependent_ on the (dialect) data stored in the metadata. I wonder whether separating the dialect information completely from the metadata, either to be retrieved separately, or left for the implementation, might not be a better approach. This comes to my question I was asking before: is the Dialect information in the metadata _prescriptive_ or _descriptive_; the current model is the former (and leads to a complexity), whereas if it was the latter, that would simplify metadata processing. (I.e., the processor would have to extend `M` with the dialect data _once the parsing occurred_ but the final merge of all the metadata could then be delayed and done only once.)

(Note that this is also related to issue #145, see https://github.com/w3c/csvw/issues/145#issuecomment-68828919, and to issue #144, see https://github.com/w3c/csvw/issues/144#issuecomment-69032693)

**Comment by gkellogg** [comment]
This is pretty close to the way I do it, but I'll not some things inline.

> # Processing steps
> 
> ## If the process starts with the URL of a CSV file:
> 1. Get the raw CSV content from the file through an `HTTP GET` (or access it from the local file system). Store the Link Header from the return, too.

I would say _open_ instead of get, even though an HTTP GET is used. _open_ implies to me that the file may be streamed, where _get_ implies that it is received all at once.

> 1. Get hold of the following metadata objects (see the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document): 
>    - metadata supplied by the user of the implementation that is processing the tabular data (`UM`) 
>    - metadata in a document linked to using a Link Header associated with the tabular data file (`LM`)
>    - file-specific metadata in a document located based on the location of the tabular data file (`FM`)
>    - directory-specific metadata in a document located based on the location of the tabular data file (`DM`)
> 2. Create a merge of these metadata `M1 = merge(UM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification

I would also insert here the default metadata, which can probably come after `DM`.

> 1. If `M1` does not include a [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions), add a default (_this still has to be defined_)

We do have a default defined for the Dialect Description:

> The default dialect description for CSV files is:
> 
> ```
> {
>   "encoding": "utf-8",
>   "lineTerminator": "\r\n",
>   "quoteChar": "\"",
>   "doubleQuote": true,
>   "skipRows": 0,
>   "header": true,
>   "headerRowCount": 1,
>   "delimiter": ",",
>   "skipColumns": 0,
>   "headerColumnCount": 0,
>   "skipBlankRows": false,
>   "skipInitialSpace": false,
>   "trim": false
> }
> ```

However, adding it to the metadata as such at the beginning is problematic, given our merge rules. You don't want the default metadata to take priority over that defined. This could be done by merging in the default dialect description (and metadata). However, my implementation just uses these as built-in values to use when a dialect property is requested, and not set either on the Table or TableGroup. Assigning it the default metadata, and effectively merging it after `DM` would have this effect, though. It would be included in, otherwise empty, TableGroup.

> 1. Possibly modify the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` if the `;header=absent` is set in the `Content-type` of the returned Link Header (see the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#headers)) 

This can be treated as creating `UM` if it is not otherwise specified. This could also cause `delimiter` to be set to `"<TAB>"` if the content type is `text/tsv`.

> 1. Parse the CSV data using the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` . An informal description on the effects of the various dialect parameter and the parsing itself is part of the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#parsing).

Yes, in particular, my CSV library can use many of the values from the dialect description natively (not all, though). Also, the default `lineTerminator` is a regular expression: `/\r?\n/`, which is more flexible about detecting a possible `<CR>` without being strict about it. Not sure we want to do this in the doc, though.

> 1. Retrieve the metadata embedded within the tabular data file (`EM`) (see again the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document). Note that this specification does not define the syntax of that embedded metadata, if any, except if `EM` consists of a single header row (providing `title` values for the corresponding `column` objects)
> 2. Determine the default Metadata (`DM`) (_this still has to be formally defined and described_)

As noted, I think this is done earlier, and is necessary to establish defaults for the Dialect Description. (Also, you're using `DM` twice, once for default another for directory. Perhaps `DD` for "Dialect Description").

> 1. Create a merge of `M = merge(DM, UM, EM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification `M = merge(DM, UM, EM, LM, FM, DM)`
> 2. The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows according to the Dialect Description, with `M` as the final metadata.

Also, as the merged metadata describes a TableGroup, the tabular data uses the Table description having the `url` the same as the tabular data location.

> 1. If required by the user, use the raw data and `M` to generate either JSON or RDF
> 
> ## If the process starts with the URL of a Metadata file:
> 
> T.B.D.

This is basically the following for me:
1. Parse the Metadata file as `UM`.
2. For each Table defined in `UM`, follow the steps in _If the process starts with the URL of a CSV file_ using `UM` as user defined metadata and `url` as the location of the tabular data file.

This can also be described in a single step by simply looking at _Content-Type_. If it is `application/json` start as a Metadata file, otherwise treat it as tabular data (CSV or TSV).

> # Comments:
> - step no. 9 is awkward. Of course, it can be optimized by storing `merge(LM,FM,DM)` separately, but the fact that the priority order places `UM` ahead of `EM` means that such a re-run becomes necessary

Indeed, but from a spec perspective this is correct. An implementation could decide to cache the result of parsing `LM,FM,DM`, but this is an optimization which could be described in a note. For any reasonable sized CSV, it probably wouldn't be much more efficient.

> - same for DM (steps nos. 8 and 9): it is unclear where exactly in the process `DM` comes in and is also part of the re-merge.

As noted, I believe it happens at the end of the first merge, is it is not likely to include more than Dialect Description defaults and boilerplate to make it valid TableGroup metadata.

> - much of the complication comes, in my view, of the fact that step no. 6, i.e., the parsing itself, is _dependent_ on the (dialect) data stored in the metadata. I wonder whether separating the dialect information completely from the metadata, either to be retrieved separately, or left for the implementation, might not be a better approach. This comes to my question I was asking before: is the Dialect information in the metadata _prescriptive_ or _descriptive_; the current model is the former (and leads to a complexity), whereas if it was the latter, that would simplify metadata processing. (I.e., the processor would have to extend `M` with the dialect data _once the parsing occurred_ but the final merge of all the metadata could then be delayed and done only once.)

As noted, I disagree. This process seems to work well for my implementation.

> (Note that this is also related to issue #145, see https://github.com/w3c/csvw/issues/145#issuecomment-68828919, and to issue #144, see https://github.com/w3c/csvw/issues/144#issuecomment-69032693)

Presumably this goes into the model document in a new normative section, and can simplify the informative processing steps for a CSV.

**Comment by iherman** [comment]
> On 05 Feb 2015, at 20:41 , Gregg Kellogg notifications@github.com wrote:
> 
> This is pretty close to the way I do it, but I'll not some things inline.
> 
> Processing steps
> 
> If the process starts with the URL of a CSV file:
> 
>   • Get the raw CSV content from the file through an HTTP GET (or access it from the local file system). Store the Link Header from the return, too.
> I would say open instead of get, even though an HTTP GET is used. open implies to me that the file may be streamed, where get implies that it is received all at once.

True. Actually, strictly speaking, an HTTP HEAD is all we need at this point, and the open occurs only if one wants to really parse. But that becomes an implementation level optimization.

>   • Get hold of the following metadata objects (see the section on locating metadata in the tabular data model document):
>       • metadata supplied by the user of the implementation that is processing the tabular data (UM)
>       • metadata in a document linked to using a Link Header associated with the tabular data file (LM)
>       • file-specific metadata in a document located based on the location of the tabular data file (FM)
>       • directory-specific metadata in a document located based on the location of the tabular data file (DM)
>   • Create a merge of these metadata M1 = merge(UM, LM, FM, DM) using the merge algorithm defined in the Tabular Metadata specification
> I would also insert here the default metadata, which can probably come after DM.
> 
>   • If M1 does not include a Dialect Description, add a default (this still has to be defined)
> We do have a default defined for the Dialect Description:
> 
> The default dialect description for CSV files is:
> 
> {
>   "encoding": "utf-8",
>   "lineTerminator": "\r\n",
>   "quoteChar": "\"",
>   "doubleQuote": true,
>   "skipRows": 0,
>   "header": true,
>   "headerRowCount": 1,
>   "delimiter": ",",
>   "skipColumns": 0,
>   "headerColumnCount": 0,
>   "skipBlankRows": false,
>   "skipInitialSpace": false,
>   "trim": false
> }
> 
> However, adding it to the metadata as such at the beginning is problematic, given our merge rules. You don't want the default metadata to take priority over that defined. This could be done by merging in the default dialect description (and metadata). However, my implementation just uses these as built-in values to use when a dialect property is requested, and not set either on the Table or TableGroup. Assigning it the default metadata, and effectively merging it after DM would have this effect, though. It would be included in, otherwise empty, TableGroup.

Well... M1 is a throw away metadata anyway! So, again conceptually, we are not talking about a merge, just adding these defaults to M1. That being said, if these two steps are switched, and the `M1` becomes `M1 = merge(UM, LM, FM, DM, DD)`, where `DD` is the default dialect, we get to the same result.

Of course, all this is conceptual, ie, implementations may optimize this (as you do). And you are right that the Dialect can be part of the general default metadata...

>   • Possibly modify the Dialect Description of M1 if the ;header=absent is set in the Content-type of the returned Link Header (see the Tabular Model Document)
> This can be treated as creating UM if it is not otherwise specified. This could also cause delimiter to be set to "<TAB>" if the content type is text/tsv.
> 
>   • Parse the CSV data using the Dialect Description of M1 . An informal description on the effects of the various dialect parameter and the parsing itself is part of the Tabular Model Document.
> Yes, in particular, my CSV library can use many of the values from the dialect description natively (not all, though). Also, the default lineTerminator is a regular expression: /\r?\n/, which is more flexible about detecting a possible <CR> without being strict about it. Not sure we want to do this in the doc, though.
> 
>   • Retrieve the metadata embedded within the tabular data file (EM) (see again the section on locating metadata in the tabular data model document). Note that this specification does not define the syntax of that embedded metadata, if any, except if EM consists of a single header row (providing title values for the corresponding column objects)
>   • Determine the default Metadata (DM) (this still has to be formally defined and described)
> As noted, I think this is done earlier, and is necessary to establish defaults for the Dialect Description. (Also, you're using DM twice, once for default another for directory. Perhaps DD for "Dialect Description").

Ah, you are right. I will redo steps to make this proper and put it as a separate comment. Just to have it cleanly done...

>   • Create a merge of M = merge(DM, UM, EM, LM, FM, DM) using the merge algorithm defined in the Tabular Metadata specification M = merge(DM, UM, EM, LM, FM, DM)
>   • The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows according to the Dialect Description, with M as the final metadata.
> Also, as the merged metadata describes a TableGroup, the tabular data uses the Table description having the url the same as the tabular data location.

I am not sure I fully understand what you say here (although, honestly, I have not looked at table groups so far, maybe we should concentrate on one table situation to clarify all details...)

>   • If required by the user, use the raw data and M to generate either JSON or RDF
> If the process starts with the URL of a Metadata file:
> 
> T.B.D.
> 
> This is basically the following for me:
> 
>   • Parse the Metadata file as UM.
>   • For each Table defined in UM, follow the steps in If the process starts with the URL of a CSV file using UM as user defined metadata and url as the location of the tabular data file.
> This can also be described in a single step by simply looking at Content-Type. If it is application/json start as a Metadata file, otherwise treat it as tabular data (CSV or TSV).

O.k. I will document this, separately

> Comments:
> 
>   • step no. 9 is awkward. Of course, it can be optimized by storing merge(LM,FM,DM) separately, but the fact that the priority order places UM ahead of EM means that such a re-run becomes necessary
> Indeed, but from a spec perspective this is correct. An implementation could decide to cache the result of parsing LM,FM,DM, but this is an optimization which could be described in a note. For any reasonable sized CSV, it probably wouldn't be much more efficient.
> 
>   • same for DM (steps nos. 8 and 9): it is unclear where exactly in the process DM comes in and is also part of the re-merge.
> As noted, I believe it happens at the end of the first merge, is it is not likely to include more than Dialect Description defaults and boilerplate to make it valid TableGroup metadata.

Will document this separately)

>   • much of the complication comes, in my view, of the fact that step no. 6, i.e., the parsing itself, is dependent on the (dialect) data stored in the metadata. I wonder whether separating the dialect information completely from the metadata, either to be retrieved separately, or left for the implementation, might not be a better approach. This comes to my question I was asking before: is the Dialect information in the metadata prescriptive or descriptive; the current model is the former (and leads to a complexity), whereas if it was the latter, that would simplify metadata processing. (I.e., the processor would have to extend M with the dialect data once the parsing occurred but the final merge of all the metadata could then be delayed and done only once.)
> As noted, I disagree. This process seems to work well for my implementation.

Well, Gregg, life would be pretty boring if we agreed on everything:-) The fact that you implemented it is only proof that it can be done. I realize that the matter of simplicity is not a precise term, though...

I think this is typically a discussion for the F2F.

> (Note that this is also related to issue #145, see #145 (comment), and to issue #144, see #144 (comment))
> 
> Presumably this goes into the model document in a new normative section, and can simplify the informative processing steps for a CSV.

+1

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
This is a second version of the processing steps write up, taking into account the comments of @gkellogg in https://github.com/w3c/csvw/issues/193#issuecomment-73112207. To be clear, and to simplify things a bit at this point, these steps describe the steps _for a single table only_, the `TableGroup` is not (yet) taken into account.

## Processing steps

### If the process starts with the URL of a CSV file
1. Open the raw CSV content from the file or access it from the local file system (typically, this is done through HTTP, but can also be part of a streaming service). Store the Link Header from the return, too.
2. Get hold of the following metadata objects (see the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document): 
   - metadata supplied by the user of the implementation that is processing the tabular data (`UM`) 
   - metadata in a document linked to using a Link Header associated with the tabular data file (`LM`)
   - file-specific metadata in a document located based on the location of the tabular data file (`FM`)
   - directory-specific metadata in a document located based on the location of the tabular data file (`DM`)
3. Define a separate `PM` for a predefined, a.k.a. default metadata (_reference to be added_). (Note that this predefined metadata also contains a default set of values for the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions.)) 
4. Create a merge of these metadata `M1 = merge(UM, LM, FM, DM, PM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
5. Possibly modify the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` if the `;header=absent` is set in the `Content-type` of the returned Link Header (see the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#headers)) 
6. Parse the CSV data using the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1`. An informal description on the effects of the various dialect parameter and the parsing itself is part of the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#parsing). 
7. Retrieve the metadata embedded within the tabular data file (`EM`) (see again the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document). Note that this specification does not define the syntax of that embedded metadata, if any, except if `EM` consists of a single header row (providing `title` values for the corresponding `column` objects)
8. Create the final metadata `M` by `M = merge(UM, EM, LM, FM, DM, PM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
9. The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows according to the Dialect Description, with `M` as the final metadata.
10. If required by the user, use the raw data and `M` to generate either JSON or RDF

### If the process starts with the URL of a Metadata file:
1. Open and parse the content of the URL yielding the metadata `SM`
2. Generate `UMS = merge(UM, SM)`, where the `UM` is the metadata supplied by the user of the implementation
3. Start the previous steps with the URL of the CSV file retrieved from `UMS` using the `url` property, and with `UMS` playing the role of `UM`

**Comment by gkellogg** [comment]
This looks great! A small point on 9): "possibly discarding the rows and/or columns according to the Dialect Description, ..." as we do have skipColumns as well as skipRows.

As far as I'm concerned, we can put this in the model document, but at this point, we may as well wait until the F2F.

Thanks for working this up, @iherman!

**Comment by iherman** [comment]
> On 06 Feb 2015, at 21:21 , Gregg Kellogg notifications@github.com wrote:
> 
> This looks great! A small point on 9): "possibly discarding the rows and/or columns according to the Dialect Description, ..." as we do have skipColumns as well as skipRows.
> 
> As far as I'm concerned, we can put this in the model document, but at this point, we may as well wait until the F2F.

Indeed:-)

Ivan

> Thanks for working this up, @iherman!
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
Just a food for thought for the F2F, as another possibility to possibly simplify the procedure...

As we said, the complication comes from the fact that the merging procedure must be performed twice: once partially to retrieve, essentially, the dialect, and then with all parties involved for the final processing. (Optimizations are of course  possible.)

The current model allows a "distributed" specification of the dialect. The separator character is defined in one, whereas the number of skipped rows in another. Is it really necessary to do that? What if the dialect is considered to be, sort of, "atomic". Ie, there is no need to do any merge; instead, in the first step, the first available dialect is retrieved from the series of metadata files; once one is found, the others are forgotten and the parsing may occur. (Because the default metadata closes the files to be merged, this includes the default dialect.) All this would make the description of the processing steps simpler...

Just a thought.

**Comment by danbri** [comment]
See http://www.w3.org/2015/02/12-csvw-irc#T11-51-20

**Comment by JeniT** [comment]
Discussed at Feb F2F. Resolved to make dialect descriptions atomic (not merged from separate metadata files), which should simplify the process. Also needs to include factoring in the role of metadata at the table group level, and the use of the default dialect.

**Comment by iherman** [comment]
As agreed at the F2F, here is the modified processing steps leading to an annotated metadata. Note that:
- I am not sure what the final terminology will be for merging; do we still use the term "merge"
- We had a small discussion on point 9 below; @gkellogg had a more "functional" view. However, if we want to have, conceptually, a final annotated tabular data, then, again conceptually, I believe it is cleaner to describe it as below, ie, a step "producing" the final metadata (how the implementation works is another matter). A reference is missing, however, on what those finalization steps are. I presume that, eventually, those will be somewhere in the metadata document.
- I am not sure it is worth reproducing the drawing of @JeniT of the F2F. If so, I can try to do that. 
- I am stil not sure where this text would go, ie, to which document. Once this is known, I am happy converting this into HTML and add it there.

## Processing steps

### If the process starts with the URL of a CSV file
1. Open the raw CSV content from the file or access it from the local file system (typically, this is done through HTTP, but can also be part of a streaming service). Store the Link Header from the return, too.
2. Get hold of the following metadata objects (see the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document): 
   - metadata supplied by the user of the implementation that is processing the tabular data (`UM`) 
   - metadata in a document linked to using a Link Header associated with the tabular data file (`LM`)
   - file-specific metadata in a document located based on the location of the tabular data file (`FM`)
   - directory-specific metadata in a document located based on the location of the tabular data file (`DM`)
3. Extract the _first available_ [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) from the metadata files `UM`, `LM`, `FM`, `DM` (in this order) yielding the final Dialect Description (`DD`). If none of the metadata have a Dialect Description, `DD` is set to the [default Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions).
4. Possibly modify `DD` if the `;header=absent` is set in the `Content-type` of the returned Link Header (see the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#headers)) by setting the `header` value to `false`.
5. Parse the CSV content using `DD`.
6. Retrieve the metadata embedded within the tabular data file (`EM`). Note that this specification does not define the syntax of that embedded metadata, if any, except if `EM` consists of a single header row (providing `title` values for the corresponding `column` objects).
7. Create the final metadata `M = merge(UM, EM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata. 
8. Update the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M` with that of `DD` (`DD` may have been modified if the `;header=absent` is set in the `Content-type` of the returned Link Header).
9. Finalize `M` according to the rules described in [_@@@@ reference to be added_]. (This step involves, for example, adding a `name` to each column if missing.) 
10. The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows and columns according to the Dialect Description (using the `header` or the `headerRowCount`, as well as the `skipRows` properties for rows, and `skipColumns` and `headerColumnCount` for columns), with `M` as the final metadata.

### If the process starts with the URL of a Metadata file:
1. Open and parse the content of the URL yielding the metadata `SM`
2. Generate `UMS = merge(UM, SM)`, where the `UM` is the metadata supplied by the user of the implementation
3. Process with the steps above as follows:
   - If `UMS` consists of a single Table Description (or a Table Group with a single Table Description), start the previous steps with the URL of the CSV file retrieved from `UMS` using the `url` property, and with `UMS` playing the role of `UM`. 
   - If `UMS` consists of a Table Group with several Table Descriptions then the processing steps have to be followed for each CSV file retrieved from `UMS` using the `url` properties, yielding the respective Annotated Tabular Data.

**Comment by gkellogg** [comment]
Note that the merge process always creates a TableGroup at the top level, so step 3 when processing a Metadata file can be collapsed.

Otherwise, this looks good. We've moved other such procedures into the model document, and this is probably the best place for it probably in section 3).

**Comment by iherman** [comment]
Ah. I did not check the new merge algorithm, but it does make sense to end up with TableGroups all the time. It unifies processing down the line.

I let you guys decide where the final text should go.

**Comment by iherman** [comment]
Actually, @gkellogg, what you said in https://github.com/w3c/csvw/issues/193#issuecomment-75597501 is not in line with the latest version of the metadata document. At present, what it says is:

> If the top-level object of either of the metadata files are table descriptions, these are treated as if they were table group descriptions containing a single table description (ie having a single resource property whose value is the same as the original table description).

I.e., a table-description metadata _may_ be transformed into a table-group-description metadata but _only_ if the "other" metadata is already one. If all metadata involved are table-description metadata then this does not occur, hence the necessity of step 3.

That being said, it may make sense if the normalization step defined in 4 contained a first step that would turn _all_ metadata into table-group-description format. I.e., _all_ the remaining processing, including the conversion, would rely on table group descriptions only. That would greatly simplify both the definition and the processing. If this is not yet decided, we may want to raise a new issue. (I will be on the road starting tomorrow so, please, do it if you agree with me.)

**Comment by gkellogg** [comment]
Note that the steps are not quite the same as @iherman described above. In particular there is no _finalize_ step, as this is at odds with my interpretation of processing metadata. So called _finalized_ values are provided using the `property value` descriptions from the metadata description for each property. Indeed finalizing make it difficult, if not impossible, to perform subsequent merges, as is done when the metadata file is the entry point.

Also, dialect description has to come from the first found for the particular table description associated with the tabular data file being used, as different table descriptors could reasonably use different dialect information. And when updating `M`, it needs to update just the dialect description for that table descriptor.

**Comment by iherman** [comment]
> On 25 Feb 2015, at 06:42 , Gregg Kellogg notifications@github.com wrote:
> 
> Note that the steps are not quite the same as @iherman described above. In particular there is no finalize step, as this is at odds with my interpretation of processing metadata. So called finalized values are provided using the property value descriptions from the metadata description for each property. Indeed finalizing make it difficult, if not impossible, to perform subsequent merges, as is done when the metadata file is the entry point.

I do not understand this remark. My intention is that the finalization happens at the end of all processing, when all merge operations are done, and we get to the final version of the annotated table instance. At that point I would expect, for example, that all column descriptions have a `name` (whether given by one of the metadata, whether it comes from a `title` value, or whether it is the default value: this should become immaterial at that point). Making the assumption that the metadata is, in this sense, complete, makes the description of, eg, the csv2json/csv2rdf specification far more simple, because that algorithm can simply rely on the existence of those data.

I am at the airport at the moment, so I cannot look at the final version of the document (and I will be pretty much swallowed by a conference in the coming days).

Ivan

> Also, dialect description has to come from the first found for the particular table description associated with the tabular data file being used, as different table descriptors could reasonably use different dialect information. And when updating M, it needs to update just the dialect description for that table descriptor.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
I understand that your view is that the finalized metadata is expanded with all defaults, but that has not been my approach. the `property value` mechanism deals with this, so it's not necessary to fully expand the metadata.

the `name` property value always returns something, but through an algorithmic behavior, rather than from the data being fleshed out. By keeping this binding loose, it becomes possible to merge in more metadata later on.

In the end, I think it is consistent to look at this either way, but the metadata document already is written to describe `property values`. What is more clear, and what is simpler is a matter of opinion. Of course, I'll yield to the direction of the group on this.

**Comment by iherman** [comment]
> On 25 Feb 2015, at 02:17 , Gregg Kellogg notifications@github.com wrote:
> 
> I understand that your view is that the finalized metadata is expanded with all defaults, but that has not been my approach. the property value mechanism deals with this, so it's not necessary to fully expand the metadata.
> 
> the name property value always returns something, but through an algorithmic behavior, rather than from the data being fleshed out. By keeping this binding loose, it becomes possible to merge in more metadata later on.
> 
> In the end, I think it is consistent to look at this either way, but the metadata document already is written to describe property values. What is more clear, and what is simpler is a matter of opinion. Of course, I'll yield to the direction of the group on this.

Well, you guys are the editors, and the real try if the test cases work for both views. So, at the end of the day, it is your (in plural) call; but then we have to be very careful how the conversion document will be described to use that approach.

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
My version of the csv2rdf doc uses the property value mechanism to good effect, I think. Note that the same logic applies to getting values of inherited properties.

**Comment by 6a6d74** [comment]
@gkellogg ... having read through the last few posts I'll admit to still being confused as to what the `property value` mechanism is. Can you provide an example please?

**Comment by gkellogg** [comment]
I'll add something to the document to clarify, but here are some based on a reading of the document:
- The `title` property is a natural language property. The property value of a natural language property is an object who's properties are language codes and values are an array of strings. Thus, if `title` were defined as ["GID", "Generic Identifier"] and the `@language` were "en", the property value of `title` would be `{"en": ["GID", "Generic Identifier"]}`.
- The `name` property is either the value defined in the definition of the metadata (if any), or the first property value of `title` having the same language as default language or `und`, and then percent encoded. So, if `title` were "Inventory Date", it's property value would be `{"und": ["Inventory Date"]}`. The first value from that would be "Inventory Date". Percent encoding this gives "Inventory%20Date", which would be the property value of the `name` property in this case.
- The value of the `_name` variable comes from the property value of `name` after percent-decoding. So, taking the previous example, it's value would be "Inventory Date".
- The property value of `tableSchema` in a Table, is either that defined within the table, or that from the containing Table Group, if any. (similar for dialect description).
- The property value of a URI Template Property is available when extracting cells, after performing variable substitution.
- The property value of an inherited property is derived by looking upwards in the descriptor hierarchy. So, if `aboutUrl` is defined on a Schema, when accessing from a Column, it first looks there, and then goes and looks in Schema where it finds it, and uses that value

Please suggest any specific examples that you think would help to clarify this.

**Comment by iherman** [comment]
> On 26 Feb 2015, at 11:01 , Gregg Kellogg notifications@github.com wrote:
> 
> I'll add something to the document to clarify, but here are some based on a reading of the document:
> 
>   • The title property is a natural language property. The property value of a natural language property is an object who's properties are language codes and values are an array of strings. Thus, if title were defined as ["GID", "Generic Identifier"] and the @language were "en", the property value of title would be {"en": ["GID", "Generic Identifier"]}.
>   • The name property is either the value defined in the definition of the metadata (if any), or the first property value of title having the same language as default language or und, and then percent encoded. So, if title were "Inventory Date", it's property value would be {"und": ["Inventory Date"]}. The first value from that would be "Inventory Date". Percent encoding this gives "Inventory%20Date", which would be the property value of the name property in this case.
>   • The value of the _name variable comes from the property value of name after percent-decoding. So, taking the previous example, it's value would be "Inventory Date".

"Inventory%20Date"

>   • The property value of tableSchema in a Table, is either that defined within the table, or that from the containing Table Group, if any. (similar for dialect description).
>   • The property value of a URI Template Property is available when extracting cells, after performing variable substitution.
>   • The property value of an inherited property is derived by looking upwards in the descriptor hierarchy. So, if aboutUrl is defined on a Schema, when accessing from a Column, it first looks there, and then goes and looks in Schema where it finds it, and uses that value
> Please suggest any specific examples that you think would help to clarify this.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
> > • The value of the _name variable comes from the property value of name
> > after percent-decoding. So, taking the previous example, it's value would
> > be "Inventory Date".
> 
> "Inventory%20Date"

Actually, it is percent-decoded to get "Inventory Date"

**Comment by iherman** [comment]
> On 26 Feb 2015, at 13:42 , Gregg Kellogg notifications@github.com wrote:
> 
> On Feb 26, 2015, at 10:28 AM, Ivan Herman notifications@github.com wrote:
> 
> > On 26 Feb 2015, at 11:01 , Gregg Kellogg notifications@github.com wrote:
> > 
> > I'll add something to the document to clarify, but here are some based on
> > a reading of the document:
> > 
> > • The title property is a natural language property. The property value
> > of a natural language property is an object who's properties are language
> > codes and values are an array of strings. Thus, if title were defined as
> > ["GID", "Generic Identifier"] and the @language were "en", the property
> > value of title would be {"en": ["GID", "Generic Identifier"]}.
> > • The name property is either the value defined in the definition of the
> > metadata (if any), or the first property value of title having the same
> > language as default language or und, and then percent encoded. So, if title
> > were "Inventory Date", it's property value would be {"und": ["Inventory
> > Date"]}. The first value from that would be "Inventory Date". Percent
> > encoding this gives "Inventory%20Date", which would be the property value
> > of the name property in this case.
> > • The value of the _name variable comes from the property value of name
> > after percent-decoding. So, taking the previous example, it's value would
> > be "Inventory Date".
> 
> "Inventory%20Date"
> 
> Actually, it is percent-decoded to get "Inventory Date"
> 
> Gregg

Hm. How do the software know that the user has not put the name value by hand with content encoding? Do we have the right to decode it even in that case? I am not sure

Ivan

> > • The property value of tableSchema in a Table, is either that defined
> > within the table, or that from the containing Table Group, if any. (similar
> > for dialect description).
> > • The property value of a URI Template Property is available when
> > extracting cells, after performing variable substitution.
> > • The property value of an inherited property is derived by looking
> > upwards in the descriptor hierarchy. So, if aboutUrl is defined on a
> > Schema, when accessing from a Column, it first looks there, and then goes
> > and looks in Schema where it finds it, and uses that value
> > Please suggest any specific examples that you think would help to clarify
> > this.
> > 
> > —
> > Reply to this email directly or view it on GitHub.
> 
> ---
> 
> Ivan Herman, W3C
> Digital Publishing Activity Lead
> Home: http://www.w3.org/People/Ivan/
> mobile: +31-641044153
> ORCID ID: http://orcid.org/0000-0003-0782-2704
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/w3c/csvw/issues/193#issuecomment-76235970.
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
> > > > • The value of the _name variable comes from the property value of name after percent-decoding. So, taking the previous example, it's value would be "Inventory Date".
> > > 
> > > "Inventory%20Date"
> > 
> > Actually, it is percent-decoded to get "Inventory Date"
> 
> Hm. How do the software know that the user has not put the name value by hand with content encoding? Do we have the right to decode it even in that case? I am not sure

That thought did occur to me, but this seems to lead to best result since
typically it will come from a title originally. Double URI encoding a title
would be quite unusual, and the name property is already constrained to not
include % if user-specified.

**Comment by JeniT** [comment]
@iherman @gkellogg are there any remaining issues here that we need to capture in other issues so that this can be closed when #259 is merged?

**Comment by iherman** [comment]
I am fine with closing this issue. I have added a small note on the PR #259, though.

Ivan

> On 28 Feb 2015, at 16:17 , Jeni Tennison notifications@github.com wrote:
> 
> @iherman @gkellogg are there any remaining issues here that we need to capture in other issues so that this can be closed when #259 is merged?
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704
